### ISSUE TITLE
Literal slicing of a variable-length tuple does not work

### ISSUE THREAD

**Issue by airallergy** [python/mypy#13058]
**Bug Report**

Slicing a variable-length tuple with a literal int does not reveal as a corresponding fixed-length tuple. This causes both false negative, where no error reported with incorrect unpacking, and false positive of incompatible type with say return values.

**To Reproduce**

```python
def foo(x: tuple[str, ...]) -> tuple[str, str]:
    a, b = x[:3]

    return x[:2]

```

**Expected Behavior**

`a, b = x[:3]` should give an error of too many values to unpack, `return x[:2]` should pass with no error.

**Actual Behavior**

`a, b = x[:3]` gives no error, `return x[:2]` says incompatible return value type.

**Your Environment**

- Mypy version used: 0.961
- Python version used: 3.10
- Operating system and version: macOS 12.4

**Comment by decorator-factory** [comment]
`x[:2]` could have a length of 0, 1, or 2. In Python, slicing automatically trims the range if the sequence is not large enough:
```py
>>> things = (1, 2, 3)
>>> things[:5]
(1, 2, 3)
>>> 
```
So erroring on `return x[:2]` is a true positive.

---
I think the reason mypy doesn't complain about the first one is the following: the type of `x[:3]` is inferred as `tuple[str, ...]`, not `tuple[()] | tuple[int] | tuple[int, int] | tuple[int, int, int]`:
```py
def foo(x: tuple[str, ...]) -> None:
    s = x[:3]
    reveal_type(s)  # tuple[str, ...]
    _a, _b = s
```
and `tuple[T ...]` allows unpacking into any number of variables is accepted.

This behavior is also present in Pyright, Basedpyright and Pyrefly. Maybe it could be improved for literal slices like that when the number is small (to prevent a large union type).
